aten/src/ATen/BatchedTensorImpl.h

```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include <bitset>
+
+#include <ATen/ArrayRef.h>
+#include <ATen/SmallVector.h>
+#include <ATen/Tensor.h>
+
+namespace at {
+
+// We assume this in a few other places in the codebase,
+// but there isn't a centralized definition.
+constexpr int64_t kVmapMaxTensorDims = 64;
+
+// Store this number of elements of BatchDims on the stack. Most people will
+// probably use <= 5 nested vmaps, but adjust this number as necessary.
+constexpr int64_t kBatchDimsStackSize = 5;
+
+struct BatchDim {
+  BatchDim(int64_t level, int64_t dim) : dim_(dim), level_(level) {}
+  int64_t dim() const {
+    return dim_;
+  }
+  int64_t level() const {
+    return level_;
+  }
+ private:
+  int64_t dim_;
+  int64_t level_;
+};
+
+using BatchDims = SmallVector<BatchDim, kBatchDimsStackSize>;
+using BatchDimsRef = ArrayRef<BatchDim>;
+
+struct TORCH_API BatchedTensorImpl : public c10::TensorImpl {
+  explicit BatchedTensorImpl(Tensor value, BatchDims bdims);
+
+  BatchDimsRef bdims() const { return bdims_; }
+  const Tensor& value() const { return value_; };
+
+  int64_t actualDim(int64_t dim, bool wrap_dim = true) const;
+
+  // Override a bunch of methods inherited from TensorImpl.
```

Comment:
Noooooo more overrides :(   Is there a reason why BatchTensorImpl cannot just redundantly record the size/strides of its internal tensor?

Reply:
BatchedTensorImpl can handle sizes without overriding (it can set the inherited `sizes_` vector).

All of the overrides here are safety checks and internal asserts (e.g., I think that strides(), is_contiguous() should produce an error message but I'm not 100% sure). Another way to do the override is to add a branch for `isBatched(Tensor)` [here](https://github.com/pytorch/pytorch/blob/176174a68ba2d36b9a5aaef0943421682ecc66d4/aten/src/ATen/templates/TensorBody.h#L209-L211) and [here](https://github.com/pytorch/pytorch/blob/176174a68ba2d36b9a5aaef0943421682ecc66d4/aten/src/ATen/templates/TensorBody.h#L194-L196).

I'm fine with deleting the overrides and punting modifying `strides()` and `is_contiguous()` to the future after I've done some benchmarking to make sure adding the branch doesn't affect performance too much.

Reply:
Oh it looks like you're overriding them to raise errors. OK